### PR TITLE
Add/nftx v3

### PR DESCRIPTION
**Pull request by justincavery**
## Description

Adding support for NFTX V3.

Changes from V2 include...

- no mint event (minted only)
- no random redeems (target redeem only)
- premium dutch auction for newly minted items (1 hour on sepolia, 10 hours on eth mainnet)

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Additional Notes

Include any additional information that you believe is important for the reviewers to know.

### CODE REVIEW

**Review by georgeroman — CHANGES_REQUESTED** [review]
There are some tests which are failing (the API ones):
image
Also, I get some type error in the test files:
image
image

File: packages/sdk/src/router/v6/addresses.ts [hunk 1]
```diff
@@ -238,6 +238,11 @@ export const NFTXZeroExModule: ChainIdToAddress = {
   [Network.Ethereum]: "0xcd4ebd768ccbe022274506f5e45d14fd90dc5be7",
 };
 
+export const NFTXV3Module: ChainIdToAddress = {
```
[the remark this answers is not in the document]
Comment (apoorvlathey) [reply packages/sdk/src/router/v6/addresses.ts:241]: updated

File: packages/sdk/src/nftx-v3/addresses.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { ChainIdToAddress, Network } from "../utils";
+
+export const VaultFactory: ChainIdToAddress = {
+  [Network.EthereumGoerli]: "0x1d552A0e6c2f680872C4a88b1e7def05F1858dF0",
```
[the remark this answers is not in the document]
Comment (apoorvlathey) [reply packages/sdk/src/nftx-v3/addresses.ts]: updated

File: packages/sdk/src/nftx-v3/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+dotEnvConfig();
+
+import { Interface } from "@ethersproject/abi";
+import { JsonRpcProvider } from "@ethersproject/providers";
+import { BigNumber, BigNumberish } from "@ethersproject/bignumber";
+import { Contract } from "@ethersproject/contracts";
+import { parseEther } from "@ethersproject/units";
+import { ethers } from "ethers";
+import axios from "axios";
+import { time } from "@nomicfoundation/hardhat-network-helpers";
```
[the remark this answers is not in the document]
Comment (apoorvlathey) [reply packages/sdk/src/nftx-v3/helpers.ts]: updated

File: packages/sdk/src/nftx-v3/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+  // for "buy" side
+  tokenIds?: number[]
+): Promise<{
+  price: BigNumberish;
+  executeCallData: string;
+}> => {
+  const chainId = await provider.getNetwork().then((n) => n.chainId);
+  const weth = Common.Addresses.WNative[chainId];
+
+  if (network && network.name === "tenderly") {
```
[the remark this answers is not in the document]
Comment (apoorvlathey) [reply packages/sdk/src/nftx-v3/helpers.ts]: updated

**Comment by apoorvlathey** [comment]
All tests are passing now.
For tests with API, we use Sepolia and for local tests, we use Mainnet.
```
$ hardhat test ./test/router/nftx-v3/listings-nftx-api.test.ts ./test/router/nftx-v3/offers-nftx-api.test.ts

  [ReservoirV6_0_1] NFTXV3 listings (with NFTX API routing)
    ✔ [eth][full][no-fees][reverts] (80438ms)
    ✔ [eth][full][no-fees][skip-reverts] (80938ms)
    ✔ [eth][full][fees][reverts] (79329ms)
    ✔ [eth][full][fees][skip-reverts] (79829ms)
    ✔ [eth][partial][no-fees][reverts] (50816ms)
    ✔ [eth][partial][no-fees][skip-reverts] (51040ms)
    ✔ [eth][partial][fees][reverts] (79448ms)
    ✔ [eth][partial][fees][skip-reverts] (79441ms)
    - [eth][full][fees][skip-reverts]

  [ReservoirV6_0_1] NFTX offers (with NFTX API routing)
    ✔ [full][no-fees][reverts] (33554ms)
    ✔ [full][no-fees][skip-reverts] (33501ms)
    ✔ [full][fees][reverts] (34309ms)
    ✔ [full][fees][skip-reverts] (33227ms)
    ✔ [partial][no-fees][reverts] (38168ms)
    ✔ [partial][no-fees][skip-reverts] (33409ms)
    ✔ [partial][fees][reverts] (33506ms)
    ✔ [partial][fees][skip-reverts] (35224ms)
    - [eth][full][fees][skip-reverts]


  16 passing (16m)
  2 pending

Done in 968.42s.

$ hardhat test ./test/router/nftx-v3/listings.test.ts ./test/router/nftx-v3/offers.test.ts

  [ReservoirV6_0_1] NFTXV3 listings
    ✔ [eth][single-order][full][no-fees][reverts] (21512ms)
    ✔ [eth][single-order][full][no-fees][skip-reverts] (20662ms)
    ✔ [eth][single-order][full][fees][reverts] (23284ms)
    ✔ [eth][single-order][full][fees][skip-reverts] (22680ms)
    ✔ [eth][single-order][partial][no-fees][reverts] (20642ms)
    ✔ [eth][single-order][partial][no-fees][skip-reverts] (20753ms)
    ✔ [eth][single-order][partial][fees][reverts] (21436ms)
    ✔ [eth][single-order][partial][fees][skip-reverts] (25272ms)
    ✔ [eth][multiple-orders][full][no-fees][reverts] (51652ms)
    ✔ [eth][multiple-orders][full][no-fees][skip-reverts] (43826ms)
    ✔ [eth][multiple-orders][full][fees][reverts] (101624ms)
    ✔ [eth][multiple-orders][full][fees][skip-reverts] (41969ms)
    ✔ [eth][multiple-orders][partial][no-fees][reverts] (64845ms)
    ✔ [eth][multiple-orders][partial][no-fees][skip-reverts] (41670ms)
    ✔ [eth][multiple-orders][partial][fees][reverts] (63896ms)
    ✔ [eth][multiple-orders][partial][fees][skip-reverts] (87803ms)
    - [eth][multiple-orders][full][fees][skip-reverts]

  [ReservoirV6_0_1] NFTXV3 offers
    ✔ [single-order][full][no-fees][skip-reverts] (22089ms)
    ✔ [single-order][full][no-fees][reverts] (21396ms)
    ✔ [single-order][full][fees][skip-reverts] (23306ms)
    ✔ [single-order][full][fees][reverts] (22562ms)
    ✔ [single-order][partial][no-fees][skip-reverts] (20350ms)
    ✔ [single-order][partial][no-fees][reverts] (23001ms)
    ✔ [single-order][partial][fees][skip-reverts] (19886ms)
    ✔ [single-order][partial][fees][reverts] (20683ms)
    ✔ [multiple-orders][full][no-fees][skip-reverts] (63293ms)
    ✔ [multiple-orders][full][no-fees][reverts] (64657ms)
    ✔ [multiple-orders][full][fees][skip-reverts] (85943ms)
    ✔ [multiple-orders][full][fees][reverts] (40838ms)
    ✔ [multiple-orders][partial][no-fees][skip-reverts] (40380ms)
    ✔ [multiple-orders][partial][no-fees][reverts] (39254ms)
    ✔ [multiple-orders][partial][fees][skip-reverts] (39605ms)
    ✔ [multiple-orders][partial][fees][reverts] (73013ms)
    - [eth][multiple-orders][partial][no-fees][skip-reverts]


  32 passing (24m)
  2 pending

Done in 1419.19s.
```

**Pull request merged**